subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -155,92 +158,94 @@ public DefaultJvmTestSuite(String name, DependencyFactory dependencyFactory, Con
+        // resolution.
+        implementation.withDependencies(dependencySet -> {
+            this.dependencies.getImplementation().bundle(getTestSuiteTestingFramework().map(vtf -> createDependencies(vtf.getImplementationDependencies())).orElse(Collections.emptyList()));
+        });
+        runtimeOnly.withDependencies(dependencySet -> {
+            this.dependencies.getRuntimeOnly().bundle(getTestSuiteTestingFramework().map(vtf -> createDependencies(vtf.getRuntimeOnlyDependencies())).orElse(Collections.emptyList()));
+        });
     }
 
-    private List<ExternalModuleDependency> createDependencies(List<ModuleVersionIdentifier> dependencies) {
-        return dependencies.stream().map(id -> {
-            String notation = id.getGroup() + ":" + id.getName() + ("".equals(id.getVersion()) ?  "" : (":" + id.getVersion()));
-            return dependencyFactory.create(notation);
-        }).collect(Collectors.toList());
+    private void initializeTestFramework(String name, Test task) {
+        Provider<TestFramework> mapTestingFrameworkToTestFramework = getTestSuiteTestingFramework().map(vtf -> {
+            switch (vtf.getFramework()) {
+                case JUNIT4:
+                    return new JUnitTestFramework(task, (DefaultTestFilter) task.getFilter(), false);
+                case KOTLIN_TEST: // fall-through
+                case JUNIT_JUPITER: // fall-through
+                case SPOCK:
+                    return new JUnitPlatformTestFramework((DefaultTestFilter) task.getFilter(), false);
+                case TESTNG:
+                    return new TestNGTestFramework(task, task.getClasspath(), (DefaultTestFilter) task.getFilter(), getObjectFactory());
+                default:
+                    throw new IllegalStateException("do not know how to handle " + vtf);
+            }
+        });
+
+        if (name.equals(JvmTestSuitePlugin.DEFAULT_TEST_SUITE_NAME)) {
+            // In order to maintain compatibility for the default test suite, we need to load JUnit4 from the Gradle distribution
+            // instead of including it in testImplementation.
+            task.getTestFrameworkProperty().convention(mapTestingFrameworkToTestFramework.orElse(getProviderFactory().provider(() -> new JUnitTestFramework(task, (DefaultTestFilter) task.getFilter(), true))));
+            // We can't disallow changes to the test framework yet because we need to allow the test task to be configured without the test suite
+            task.getTestFrameworkProperty().finalizeValueOnRead();
+        } else {
+            // The Test task's testing framework is derived from the test suite's testing framework
+            task.getTestFrameworkProperty().convention(mapTestingFrameworkToTestFramework);
+            // The Test task cannot override the testing framework chosen by the test suite
+            task.getTestFrameworkProperty().disallowChanges();
```

Comment:
This is the key to this whole thing, isn't it.

Reply:
Yeah, and the `finalizeValueOnRead` helps too.  That ensures we get a single instance of the mapping